### ISSUE TITLE
Training-Time updated custom Layer attributes are NOT saved

### ISSUE THREAD

**Issue by arbiter1elegantiae** [tensorflow/tensorflow#39583]
**System information**

- OS Platform and Distribution: Linux Ubuntu 19.10
- TensorFlow installed from: Docker Image latest-gpu-py3-jupyter
- TensorFlow version: v2.1.0-rc2-17-ge5bf8de 2.1.0
- Python version: (major, minor, micro, release level, serial) (3, 6, 9, 'final', 0)
- CUDA/cuDNN version: 10.1
- GPU model and memory: GeForce GTX 1050 Ti

**Describe the current behavior**
I am not able to save the updated attribute value `self.ratio` of the following custom `tf.keras.layer.Layer`
```
   class Kwta(Layer):

    def __init__(self, ratio, conv = False, data_format="channels_last", **kwargs):
        super(Kwta, self).__init__(**kwargs)
        self.conv = conv
        self.ratio = ratio
        self.data_format = data_format
        
    def build(self, input_shape):
        #Not important here, see Reproducible Example Section for the full code
        
    def call(self, inputs):
      #Not important here, see Reproducible Example Section for the full code
    
    def get_config(self):
        my_config = {
            'conv' : self.conv,
            'ratio' : self.ratio,
            'data_format' : self.data_format,
            'dim' : self.dim
        }

        base_config = super(Kwta, self).get_config()
        return dict(list(base_config.items()) + list(my_config.items()))
```
Which gets incrementally updated while training by the following custom `tf.keras.callbacks.Callback`

```
class incremental_learning_withDecreasing_ratio(tf.keras.callbacks.Callback):
    """ Icrementally adjust the Kwta self.ratio attribute every 2 epochs.
         End the learning process when ratio == 0.3  """

    def __init__(self, delta = 0.05):
        super(incremental_learning_withDecreasing_ratio, self).__init__()
        self.delta = delta

    def on_epoch_begin(self, epoch, logs=None):
        # The update occurs at the beginning of every 2 epochs
        if epoch % 2 == 0:  
            for i in range(1, 5): # For each Kwta layer
                name = 'kwta_'+str(i)
                layer = self.model.get_layer(name = name)      
                layer.ratio -= self.delta
            
            print('\n Fine tuning: current ratio {} \n'.format(round(layer.ratio, 2)))
    
    def on_epoch_end(self, epoch, logs=None):
        layer = self.model.get_layer('kwta_1')
        if ( round(layer.ratio, 2) == 0.3 ) and epoch % 2 == 1: 
                print('\n Desired Ratio reached, stop training...')
                self.model.stop_training = True
```
This is the whole model on which I use the Kwta layer and the callback

```
kwta_cnn = tf.keras.models.Sequential([
  layers.Conv2D(32, 3, padding='same', activation=None, input_shape = (32, 32, 3)),
  Kwta(ratio=0.6, conv=True, name='kwta_1'),
  layers.Conv2D(32, 3, padding='same', activation=None),
  Kwta(ratio=0.6, conv=True, name='kwta_2'),
  layers.MaxPooling2D(pool_size=(2,2)),
  layers.Dropout(0.2, seed=42),

  layers.Conv2D(64, 3, padding='same', activation=None),
  Kwta(ratio=0.6, conv=True, name='kwta_3'),
  layers.Conv2D(64, 3, padding='same', activation=None),
  Kwta(ratio=0.6, conv=True, name='kwta_4'),
  layers.MaxPooling2D(pool_size=(2,2)),
  layers.Dropout(0.3, seed=42),

  layers.Flatten(),
  layers.Dense(10, activation='softmax', kernel_regularizer= tf.keras.regularizers.l2(0.0005))
])
```
When I fit the model `self.ratio` is correctly updated until its value is `0.3` as printed by the callback
```
history_kwta_ft = kwta_cnn.fit(x=x_train, y=y_train, epochs = 20, callbacks=[incremental_learning_withDecreasing_ratio()])
```
```
......
 
Fine tuning: current ratio 0.4 

Epoch 7/20
50000/50000 [==============================] - 40s 798us/sample - loss: 0.9049 - accuracy: 0.7080
Epoch 8/20
50000/50000 [==============================] - 39s 787us/sample - loss: 0.8921 - accuracy: 0.7122

 Fine tuning: current ratio 0.35 

Epoch 9/20
50000/50000 [==============================] - 39s 782us/sample - loss: 0.8862 - accuracy: 0.7141
Epoch 10/20
50000/50000 [==============================] - 39s 785us/sample - loss: 0.8819 - accuracy: 0.7147

 Fine tuning: current ratio 0.3 

Epoch 11/20
50000/50000 [==============================] - 40s 793us/sample - loss: 0.8812 - accuracy: 0.7174
Epoch 12/20
49920/50000 [============================>.] - ETA: 0s - loss: 0.8651 - accuracy: 0.7218
 Desired Ratio reached, stop training...
```
**Here is where the issue emerges**: I am getting two way different accuracies when I evaluate the trained model on the test set before and after saving it:
```
# Before Saving
kwta_cnn.evaluate(x = x_test, y = y_test, batch_size=128)
10000/10000 [==============================] - 5s 534us/sample - loss: 0.9503 - accuracy: 0.7017

kwta_cnn.save('saved_custom_model/kwta_cnn_minimal') 
del kwta_cnn
kwta_cnn = tf.keras.models.load_model('saved_custom_model/kwta_cnn_minimal')

# After Saving
kwta_cnn.evaluate(x = x_test, y = y_test, batch_size=128)
10000/10000 [==============================] - 3s 341us/sample - loss: 1.1248 - accuracy: 0.6349
```
After debugging for a while, I am strongly confident this drop in accuracy is due to `self.ratio` not being correctly saved by `model.save()`. In particular, I believe TF only store the original attribute (in this specific case `0.6`) and cannot deal with dynamic updated attributes that are neither `tf.Variable()` nor `self.add_weight()`. Unfortunatly the latter is not an option in my case since I seem to be able to handle only integers inside the `call(inputs)` method to perform my computation and not `tf.Tensor`s.

**Describe the expected behavior**
To be able to save and store a `tf.keras.layers.Layer` with training-time updated attributes.  

**Standalone code to reproduce the issue**
 [Colab Minimal Example](https://colab.research.google.com/drive/1LHvG8yNrX3C9SJHJ2OojZCryFSCw8nG0?usp=sharing)

**Comment by Saduf2019** [comment]
I am able to replicate the gist shared above, please find the replicated [gist here](https://colab.sandbox.google.com/gist/Saduf2019/73f7d2b339de0f2c29d5923c8190b469/untitled182.ipynb)

**Comment by arbiter1elegantiae** [comment]
After further debugging I figured out that the attribute is actually very well stored by `model.save` it was just me missing to add the label `custom_objects = {'Kwta':Kwta}` while loading the model. No I am able to access it.

So what is causing the drop in Accuracy after saving the model? What is even more strange is that the predictions before and after saving (resp. model names `kwta_cnn`, `kwta_cnn_after`) are actually the same! 

```
tf.reduce_sum(tf.cast(tf.math.logical_not(tf.math.equal(kwta_cnn.predict(x_test), kwta_cnn_after.predict(x_test))), dtype=tf.float32))

<tf.Tensor: shape=(), dtype=float32, numpy=0.0>
```
Therefore I believe the problem has to do with `model.Evaluate()` which somehow does not update the computation in `call()` when it relies on non-tensors, sticking with the original value of the attribute, not the one updated by the callback. Thus I experience such different evaluations, because before saving the original value of `self.ratio=0.6` and after saving will be `self.ratio=02`.
At least this is my intuition. It would be nice if someone could point me out what is really going on.

However, I have managed to correctly get same Accuracy before and after saving by defining the attribute `self.ratio` as a `tf.Variable(ratio, traiable=False)` instead of a float.

**Comment by sushreebarsa** [comment]
Was able replicate the issue with TF v2.5,please find the gist[ here](https://colab.research.google.com/gist/sushreebarsa/7413aa8d0f29dc7bb7583d9a14df7405/untitled182.ipynb#scrollTo=74S0CjIiW61I)....Thanks!

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
